apps/laboratory/src/hooks/useWalletConnectCosigner.ts

```diff
@@ -0,0 +1,238 @@
+import axios, { AxiosError } from 'axios'
+import { bigIntReplacer } from '../utils/CommonUtils'
+import type { UserOperation } from 'permissionless'
+
+// Define types for the request and response
+type AddPermission = {
+  permissionType: string
+  data: string
+  required: boolean
+  onChainValidated: boolean
+}
+
+type AddPermissionRequest = {
+  permission: AddPermission
+}
+
+export type AddPermissionResponse = {
+  pci: string
+  key: string
+}
+
+type Signer = {
+  type: string
+  data: {
+    ids: string[]
+  }
+}
+
+type SignerData = {
+  userOpBuilder: string
+}
+
+type PermissionsContext = {
+  signer: Signer
+  expiry: number
+  signerData: SignerData
+  factory?: string
+  factoryData?: string
+  permissionsContext: string
+}
+
+type UpdatePermissionsContextRequest = {
+  pci: string
+  signature?: string
+  context: PermissionsContext
+}
+
+type RevokePermissionRequest = {
+  pci: string
+  signature: string
+}
+
+type CoSignRequest = {
+  pci: string
+  userOp: UserOperation<'v0.7'>
+}
+
+type CoSignResponse = {
+  userOperationTxHash: string
+}
+
+// Define a custom error type
+export class CoSignerApiError extends Error {
+  constructor(
+    public status: number,
+    message: string
+  ) {
+    super(message)
+    this.name = 'CoSignerApiError'
+  }
+}
+
+export function useWalletConnectCosigner() {
```

Comment:
address and projectId are repeated all along, would it make sense to make them params and state vars instead of requiring them in all methiods?

Reply:
I think it better to accept these params on methods for now, will refactor it if structure changes later(expected).

Reply:
> I think it better to accept these params on methods for now
Why? Its fine sharing an opinion but we need to explain the reasoning behind it.